lib/config_verify/util.go

```diff
@@ -0,0 +1,292 @@
+	if !valid {
+		return valid, msgPrefix + " must be one of: " + strings.Join(allowedValues, ", ")
+	}
+	return valid, ""
+}
+
+func checkIPAddress(ip string) error {
+	if net.ParseIP(ip) == nil {
+		return errors.New("ip Address is invalid")
+	}
+	return nil
+}
+
+func extractIPsFromCertsByIP(certsByIp []config_parser.CertByIP) []string {
+	ips := []string{}
+	for _, el := range certsByIp {
+		ips = append(ips, el.IP)
+	}
+	return ips
+}
+
+func checkCertValid(keys []keydetails) *list.List {
+	errorList := list.New()
+	for _, el := range keys {
+		block, _ := pem.Decode([]byte(el.key))
+		if block == nil {
+			errorList.PushBack("Invalid format. Failed to decode " + el.certtype + " for " + el.svc)
+		} else if el.certtype == rootCa {
+			rootCA, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				errorList.PushBack("Failed to parse " + el.certtype + " for " + el.svc)
+			} else {
+				if time.Now().After(rootCA.NotAfter) {
+					errorList.PushBack(el.certtype + " for " + el.svc + " certificate has expired.")
+				}
+			}
+		}
+	}
+	return errorList
+}
+
+func validateCerts(config *config_parser.HAOnPremConfigToml) *list.List {
+	errorList := list.New()
+
+	if config.Automate.Config.EnableCustomCerts {
+		validateAutomateCerts(config, errorList)
+	}
+
+	if config.ChefServer.Config.EnableCustomCerts {
+		validateChefServerCerts(config, errorList)
+	}
+
+	if config.Postgresql.Config.EnableCustomCerts {
+		validatePostgresqlCerts(config, errorList)
+	}
+
+	if config.Opensearch.Config.EnableCustomCerts {
+		validateOpensearchCerts(config, errorList)
+	}
+
+	return errorList
+}
+
+func validateAutomateCerts(config *config_parser.HAOnPremConfigToml, errorList *list.List) {
+	if len(config.Automate.Config.CertsByIP) > 0 {
+		if len(strings.TrimSpace(config.Automate.Config.RootCA)) > 0 {
+			errorList.PushBackList(checkCertValid([]keydetails{
+				{key: config.Automate.Config.RootCA, certtype: rootCa, svc: automate},
+			}))
+		}
+
+		if !stringutils.SubSlice(config.ExistingInfra.Config.AutomatePrivateIps, extractIPsFromCertsByIP(config.Automate.Config.CertsByIP)) {
+			errorList.PushBack("Missing certificates for some automate private ips. Please make sure certificates for the following ips are provided in certs_by_ip: " + strings.Join(config.ExistingInfra.Config.AutomatePrivateIps, ", "))
+		}
+
+		for _, a2Node := range config.Automate.Config.CertsByIP {
+			if len(strings.TrimSpace(a2Node.IP)) < 1 ||
+				len(strings.TrimSpace(a2Node.PrivateKey)) < 1 ||
+				len(strings.TrimSpace(a2Node.PublicKey)) < 1 {
+				errorList.PushBack("Field certs_by_ip for Automate requires ip, private_key and public_key. Some of them are missing.")
+			}
+
+			errorList.PushBackList(checkCertValid([]keydetails{
+				{key: a2Node.PrivateKey, certtype: privateKey, svc: automate + " cert_by_ip for ip " + a2Node.IP},
+				{key: a2Node.PublicKey, certtype: publicKey, svc: automate + " cert_by_ip for ip " + a2Node.IP},
+			}))
+		}
+	} else {
```

Comment:
No test for the else case scenario, the same for other types validation methods as well

Reply:
had written test cases for failed and success scenarios.

Reply:
Updated code.